onnxruntime/core/framework/bfc_arena.cc

```diff
@@ -407,6 +415,88 @@ void BFCArena::Free(void* p) {
   }
 }
 
+Status BFCArena::Shrink() {
+  std::lock_guard<OrtMutex> lock(lock_);
+  auto num_regions = region_manager_.regions().size();
+  std::vector<void*> region_ptrs;
+  std::vector<size_t> region_sizes;
+  region_ptrs.reserve(num_regions);
+  region_sizes.reserve(num_regions);
+
+  // Even if any byte is left unused in the first allocation region, we do not want to consider it for de-allocation
+  // We never want to shrink the initial allocation if the arena extend strategy is kNextPowerOfTwo.
+  // This could seem confusingly arbitrary but the rationale is as follows:
+  // The user selected initial allocation chunk is only valid for the arena extend strategy kNextPowerOfTwo
+  // and the user has likely chosen this initial value so that any ad-hoc arena extensions/shrinkages could potentially
+  // be avoided. So we do not consider the initial allocation for shrinkage whatever its usage status.
+  // On the other hand, if the arena extension strategy is kSameAsRequested, any initial chunk set by the user or otherwise,
```

Comment:
Not quite following why I couldn't set an initial size that was meaningful and provided the majority of the required memory, and use kSameAsRequested on top of that to minimize growth beyond that. 

Or alternatively, what do we gain by adding the complication of treating kSameAsRequested differently to kNextPowerOfTwo?

Reply:
I think I understand now. If kSameAsRequested we don't use the initial size do we? However that seems slightly limiting and inconsistent. 

Would it be better to, regardless of growth strategy, always make the first allocation using the initiali size value so things are simpler?

Reply:
Even when using kNextPowerOfTwo, the initial chunk request may not be honored. Consider if the first memory request being greater than the initial chunk set, then it will go with the rounded value of the memory request itself (ignoring initial chunk set). So, initial chunk is more of a hint. to the allocator...

Reply:
Can we make things a bit more consistent though? I'm thinking something along the lines of:
  - first allocation is always the larger of initial chunk size and requested size
    - i.e. this behavior is not dependent on the growth strategy (as it's the initial allocation not the growth)
  - we always preserve the first region, whether that was sized based on the initial chunk size or the first request size is ignored
    - premise is there's no benefit in shrinking to zero, so why have complexity around the growth strategy in the Shrink code
    - we can refine later if there are real edge cases where this matters (I can think of a theoretical one, but not sure it will ever happen in the real world)

Reply:
This is  good suggestion but can we do this in a different PR ? For the arena extend strategy `kSameAsRequested`, this is a significant change in behavior now and I think it is worth deliberating a bit more. I am also not sure if it is in the scope of this PR to alter existing arena extension behavior for any of the available strategies. All this PR is trying to do is provide a mechanism to shrink the arena...